refactor(Elevator): implement state pattern

braedenkloke opened the pull request:
## What type of PR is this? (check all applicable)

- [ ] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] ✅ Test
- [X] 🧑‍💻 Refactor

## Description

Stupendous implementation of the legendary state pattern for class Elevator.

Some tasks completed:
- added package StatePatternLib with classes Context and State
- added elevator states as subclasses of Elevator. This allows various elevator states to access private methods in Elevator.
- removed deprecated method Elevator.processMessage

## Added/updated tests?

- [ ] Yes
- [X] No, and this is why: Cause. Main() runs. No side effects.
- [ ] I need help with writing tests

## Screenshots

Will Ferrell from Elf in that elevator with all the buttons.

braedenkloke committed:
feat(StatePatternLib): added initial abstract classes for Context and State

Both abstract classes are not final and should be expanded on.
To promote incremental development and testing, both classes were implemented with the
minimal  methods and attributes.

braedenkloke committed:
feat(Elevator): implement Elevator as subclass of Context

Simply made Elevator extend Context. Added super() invocation to Elevator constructor with intial state set to NULL.

Main() runs. No side effects.

braedenkloke committed:
feat(ElevatorSubsystem): add receiving state to elevator

Add ReceivingState to package ElevatorSystem.States. Decided to keep States in each subsystem rather than at the StatePatternLib level since they are specfic to each subsystem state machine. In other words, sub systems should not share states (even though they look the same!).

Main() runs. No side effects.

braedenkloke committed:
feat(ReceivingState): implement method entry to send state update

Additional work includes:
- refactoring ReceivingState to be a subclass of Elevator
- adding default constructor to Context
- setting intial state of state machine in method run

Main() runs. No side effects.

braedenkloke committed:
feat(MovingState): handle MoveElevatorCommand

State machines correctly handles MoveElevatorCommand.
Transitions from ReceivingState to MovingState and back to ReceivingState.
Additional work includes:
- add superclass ElevatorState
- add method State.doActivity

Main() runs. No side effects.

braedenkloke committed:
feat(LoadingState): handle MovePassengersCommand

Main() runs. No side effects.

braedenkloke committed:
refactor(Elevator): remove deprecated method processMessage

Message processing is being handle by state machine now.

Main() runs. No side effects.

braedenkloke committed:
refactor(LoadingState): move unload action to method entry

braedenkloke committed:
docs: update javadocs

braedenkloke committed:
docs: update javadocs

ArthurAtangana commented on src/Subsystem/ElevatorSubsytem/Elevator.java:
```diff
@@ -108,33 +115,131 @@ private void sendStateUpdate(){
+            } else if (event instanceof MovePassengersCommand) {
+                ((ElevatorState) state).handleMovePassengersCommand();
+            } else {
+                InvalidTypeException e = new InvalidTypeException("Event type received cannot be handled by this subsystem.");
```
the elevator should do any processing in the elevator other than basically transition between states. this run method shouldnt have the logic of processing events. this should be in the receiving state.

ArthurAtangana replied:
having a runstate() method in the states where it runs entry, doactivity,exit would take care of all the logic. the elevator would only be responsible of starting the runstate for each state, delegating the actual processing to the state as much as possible.

braedenkloke replied:
I'm sorry, I disagree with comment 1. 

If we are attempting to implement Elevator as a State Machine, then run() SHOULD have the logic of *handling* events. My implementation doesn't *process* the events. Processing is done in the various handler methods (eg. handleMoveElevatorCommand)

Addressing comment 2. Method changeState() is synonymous with runState().

ArthurAtangana commented on src/StatePatternLib/State.java:
```diff
@@ -0,0 +1,38 @@
+    public void doActivity() {
+        // Would have liked to name method 'do'. But 'do' is a reserved keyword in Java.
+    }
+}
```
if we are to use this pattern, there should be an exit() method.

ArthurAtangana replied:
there should be a runState() method or similar, that calls entry, doactivity and exit sequencially, see other comments

braedenkloke replied:
Addressing comment 1: Added method exit.

Addressing comment 2: method changeState is synonymous to runState.

ArthurAtangana commented on src/Subsystem/ElevatorSubsytem/Elevator.java:
```diff
@@ -108,33 +115,131 @@ private void sendStateUpdate(){
+     * @author Braeden Kloke
+     * @version March 5, 2024
+     */
+    public class ReceivingState extends ElevatorState {
```
factor out of the file

ArthurAtangana replied:
this is where we should wait to receive the message, not in the elevator run(). basically in the entry method, we do the receiving method, then in the do activity we process it and select the next state

braedenkloke replied:
Comment 1: Refactored. Done.

Comment 2: I don't agree with the suggested implementation. See bulleted list below for reasoning.

- Events act upon a state machine. In other words, an event happens and whatever state we are in has to handle it. 
- Events happen by being dequeued from the receiver.
- After an event happens, the current state in our state machine needs to immediately respond to the event.
- Thus, encapsulating 'events happening' all inside the ReceivingState would not be implementing a state machine because this implies events ONLY happen in the receiving state. This is false since events should be able to happen in any state.
- My implementation is more representative of an actual state machine where events have the possibility of happening in any state.

mpdesantis replied:
I think I'm understanding the debate. The way I see it, Braeden has the Elevator thread doing the initial processing of the message popped off of the Receiver (ie. switching it as either the MoveElevatorCommand or MovePassengersCommand), and then calling the ReceivingState's method which then transitions to the appropriate state. I think Arthur is suggesting just having the run() method basically just set an initial state, and then only having the ReceivingState capable of popping/processing commands.

I think if we put the process message logic in the ReceivingState, that seems to make sense to me. I agree with Braeden's points about the state machine and that events can happen to it regardless of what state it is in; however, if a state has no transition for a certain event, nothing needs to happen. Again, I don't know if I'm understanding it right, but I think it is actually desired behaviour that only the ReceivingState is capable of getting the MoveElevatorCommand/MovePassengersCommand. Other events still happen in all the other states (eg. time-triggered event of load-time elapsing in load() -> like the time-to-live  TL=LOAD_TIME thing in Wainer's extended FSMs, if you remember that!; same thing with the move() command and TL=TRAVEL_TIME). 

For me, I was confused thinking that MoveElevatorCommand/MovePassengersCommand were the only events here; they're just the only events that are relevant when in the ReceivingState. If we're in a non-ReceivingState, I don't think we want to be popping anything off the queue - we just want our time-triggered event to fire, move us back to the ReceivingState, and then pop off queue.

So, I think the functionality is basically going to be the same either way, but yeah - maybe makes sense to only have MoveElevatorCommand/MovePassengersCommand pop/split in the ReceivingState. Might also be more in line with what we'll be doing for the other state machines.

P.S. Kindly disregard if this makes no sense.

ArthurAtangana commented on src/Subsystem/ElevatorSubsytem/Elevator.java:
```diff
@@ -108,33 +115,131 @@ private void sendStateUpdate(){
+     * @author Braeden Kloke
+     * @version March 5, 2024
+     */
+    public class LoadingState extends ElevatorState {
```
factor out of the file

braedenkloke committed:
refactor(Elevator): move states to package level, change various methods to package private

Temporarily changed attribute in context to public to make code run.
Plan to write a getter method to change attribute event back to protected access modifier.

Main() runs. No side effects.

braedenkloke committed:
refactor(Context): change method name FROM setState TO changeState

braedenkloke committed:
refactor(Context.changeState): change arg name FROM state TO nextState

braedenkloke committed:
feat(State): add method exit

Additional refactored Context changing various this.attribute to just attribute.

Main() runs. No side effects.

braedenkloke committed:
fix(Context): make attribute event FROM public TO protected

Added method getEvent to get most recent event acting on the system.

Main() runs. No side effects.

braedenkloke committed:
docs(Context): add explaination for include event as a attribute

braedenkloke commented:
@ArthurAtangana I've addressed your comments and I've included comments for all the decisions behind my implementation of a state machine. Ready for review.

mpdesantis reviewed (approved):
Looks good, actually runs :) Thanks buddy.
